src/providers.rs

```diff
@@ -235,12 +235,12 @@ pub fn do_unregister_provider(caller: Principal, provider_id: u64) -> bool {
 }
 
 /// Changes provider details. The caller must be the owner of the provider.
-pub fn do_update_provider(caller: Principal, args: UpdateProviderArgs) {
+pub fn do_update_provider(caller: Principal, is_controller: bool, args: UpdateProviderArgs) {
     PROVIDERS.with(|providers| {
         let mut providers = providers.borrow_mut();
         match providers.get(&args.provider_id) {
             Some(mut provider) => {
-                if provider.owner != caller {
+                if !(provider.owner == caller || is_controller) {
```

Comment:
The `trap` message needs to be updated. Shouldn't it simply be "You are not authorized" as above?

Reply:
You're right; thanks. Updated. Trying to do too many things at the same time, haha.